### PR TITLE
Add HTTP API to scheduler

### LINKED ISSUE #5935: Add REST API to scheduler
I've had a couple of independent conversations lately where folks want to interact with the scheduler from some external service. The specific use cases were external process managers that can scale Dask clusters up and down. Scaling down gracefully requires calling the `retire_workers` and `workers_to_close ` methods on the scheduler RPC.

Using the RPC for this is problematic because success is heavily dependent on the `dask`, `distributed` and `python` versions used by the scheduler and the external manager. Mismatches can result in failure.

A workaround for this is exposing those methods via a RESTful endpoint. This would allow for a wider range of versions to be supported and means the external process manager doesn't even have to be written in Python.

In a conversation with @stephan-erb-by and @philipp-sontag-by around the Kubernetes operator in dask/dask-kubernetes#256 they mentioned they had done this via a scheduler plugin that added extra HTTP routes to the Dashboard web server. This is fine but does require a plugin to be installed for all distributed clusters managed by the external process manager (the operator in this case).

I wanted to open this issue to gauge the feeling of adding a more official REST API to the scheduler that exposes some general RPC methods via HTTP in a language-agnostic way.

### PR DESCRIPTION
**Pull request by Matt711**
Closes #5935 

This PR exposes some of the scheduler endpoints we need to replace the Dask RPC in the [Operator](https://github.com/dask/dask-kubernetes/issues/256).

- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by quasiben** [comment]
add to allowlis

**Comment by quasiben** [comment]
add to allowlist

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/api.py:37]: We should set the `Content-Type` header here to `text/json` and `json.dumps` this dict.

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/api.py:47]: Same here this should be a JSON response.

File: distributed/http/scheduler/tests/test_scheduler_http.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/tests/test_scheduler_http.py:273]: We should check that the workers actually retire. And also that we get the expected JSON response.

File: distributed/http/scheduler/tests/test_scheduler_http.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/tests/test_scheduler_http.py:282]: We should asset that this has a JSON header.

File: distributed/http/scheduler/tests/test_scheduler_http.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/tests/test_scheduler_http.py:294]: We should assert this has a JSON header.

File: distributed/http/scheduler/tests/test_scheduler_http.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/tests/test_scheduler_http.py:273]: And that is has a JSON header.

File: distributed/http/scheduler/tests/test_scheduler_http.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/tests/test_scheduler_http.py:257]: In the other tests you are using `aiohttp`. Perhaps it would be good to use that here for consistency.

**Review by jacobtomlinson — COMMENTED** [review]
This is looking great! It's exactly what we need for external resource managers to interact with the scheduler.

We should definitely add a documentation page about this API too.

**Comment by Matt711** [comment]
> This is looking great! It's exactly what we need for external resource managers to interact with the scheduler.
> 
> We should definitely add a documentation page about this API too.

Great! I can create a follow up PR for more documentation

**Review by jacobtomlinson — APPROVED** [review]
This looks good to me.

@fjetter would you mind reviewing here? Given you were involved in the early conversations about this it would be good to check you are happy with this approach.

**Review by jacobtomlinson — CHANGES_REQUESTED** [review]
Actually thinking a bit more we don't have an implementation here for `workers_to_close`. 

We have `retire_workers` which allows you to retire by name, but we don't have a way or retiring `n` workers. Exposing `workers_to_close` would allow that.

https://github.com/dask/distributed/blob/77cfc7304233d4558b724fd8ed93172d64e230c4/distributed/scheduler.py#L5829

File: docs/source/http_services.rst
Comment (jacobtomlinson) [inline docs/source/http_services.rst:60]:
```suggestion
.. code-block:: json

    {
```

File: docs/source/http_services.rst
Comment (jacobtomlinson) [inline docs/source/http_services.rst:67]:
```suggestion
.. code-block:: json

    {
```

**Review by jacobtomlinson — APPROVED** [review]
Thanks for the quick turnaround. I'm happy with this as a good foundation we can build on.

**Comment by jacobtomlinson** [comment]
I intend to merge this in 24 hours if there are no further comments.

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:22]: I would strongly encourage to redefine the interface of the API call and not just pass kwargs through. A primary goal of this API was to offer a more stable API and I'm not willing to couple the function signature to the HTTP API stability

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:15]: In a true REST API we would not wait for the retirement to finish but the POST would create a background job and return a UUID.
A subsequent GET would allow us to check whether that job completed already. This way the HTTP request does not block. Are we willing to go that far?

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:28]: Why is this a POST? It doesn't modify any resources, does it?

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:36]: a generic exception is not a 400 but rather a 500. A 400 would indicate wrong parameters, etc. but a generic exception might be anything (e.g. serialization error of the workers_to_close dictionary)

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:37]: This print should either be removed or replaced with a log

**Review by fjetter — COMMENTED** [review]
I think there are a few cosmetic questions around how we want to deal with exceptions but this should not block the PR. The more fundamental question is how we want to maintain API stability. Merely passing through kwargs is not what I had in mind when we first discussed this API

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:22]: Fair enough. Do you have any particular issue with the expected parameters or is this just an issue with coupling? 

I don't see any reason to make the API different to the underlying function, but we could definitely decouple it so that if the underlying function changes the API does not.

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:15]: We weren't aiming for REST here, more of an HTTP RPC. There are a lot of things un-REST-ful about this design. 

But I would be ok with returning early and adding another method to poll for completion. What do you think?

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:28]: This closes workers by name, which is a modification.

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:36]: Agree this should be 500.

File: distributed/http/scheduler/api.py
Comment (fjetter) [reply distributed/http/scheduler/api.py:28]: It only provides the worker to be closed but doesn't close any of them

Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:28]: Oh yeah I see. I misunderstood what that function was doing.

@Matt711 I don't think we actually need to expose this method at all. `retire_workers` does everything we need.

Comment (Matt711) [reply distributed/http/scheduler/api.py:28]: Since we're planning on using this API in the scheduler, don't we need to know which workers we close before closing them? Because we're going to have to close the Kubernetes pods? @jacobtomlinson

File: distributed/http/scheduler/api.py
Comment (fjetter) [reply distributed/http/scheduler/api.py:22]: I do see reason to not even allow all arguments to be passed here. For instance, the kwarg `remove` is something I would consider to be an internal-use-only argument. I don't see a valid use case for any user to control this behavior. Same for `stimulus_id` which should arguably set by the request handler, not the user.

File: distributed/http/scheduler/api.py
Comment (fjetter) [inline distributed/http/scheduler/api.py:37]: 1. This is not a `Bad request` but an internal server error now.
2. Why are we implementing an exception handler for this? Why not for all the request handlers? Isn't this default / 500 exception handler handled by `RequestHandler`, by default?
3. What happens to all other handlers if an exception occurs?

Comment (Matt711) [reply distributed/http/scheduler/api.py:37]: Because it's taking arguments, and I wanted to handle that. So I guess we should include exception handling for all request handlers?

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [reply distributed/http/scheduler/api.py:28]: It's fine, we should be able to use `retire_workers` which reports the names of the workers it retires. Then we can just clean up the completed pods after they exit.

File: distributed/http/scheduler/api.py
Comment (jacobtomlinson) [inline distributed/http/scheduler/api.py:21]: We probably want to support retiring a number of workers too rather than just by name.

File: docs/source/http_services.rst
Comment (fjetter) [inline docs/source/http_services.rst:66]: I think it would be helpful to document an example response

File: docs/source/http_services.rst
Comment (fjetter) [inline docs/source/http_services.rst:57]: It would be nice if this mentioned specifically that _addresses_ are expected and not names. Sometimes they can be interchangable but this API expects addresses. The example below might still be misleading since we're setting names, by default, to the addresses.

**Review by fjetter — COMMENTED** [review]
Just a few nits about documentation (this is intended to be public API for users, therefore I think we should invest a bit in documentation even though the API is simple)

@jacobtomlinson I would trust you with this. If you think the additional docs are not necessary, feel free to merge. These comments should not necessarily block anybody.

**Comment by jacobtomlinson** [comment]
We are blocked on this PR for some work we want to finish this week, so I think I'll just ask @Matt711 to make a follow up to improve the docs if that's ok. Totally agree with the comments though.

**Comment by philipp-sontag-by** [comment]
Hey I just saw this PR.
My current understanding is that there are two modes for the `/api/v1/retire_workers` route:

1.  Retire `n` workers based on the `workers_to_close` method of the scheduler
2. Retire the provided workers based on their address

But currently there is no API that exposes the `workers_to_close` method directly. 
At least in our use case that would be really helpful though since our graceful shutdown implementation has two phases:

1. Retrieve workers to close from the scheduler and do some custom graceful shutdown logic
2. Actually make the `retire_worker` call to the scheduler. 

@jacobtomlinson/@Matt711 Do you think that this might also be interesting for you?

**Comment by jakirkham** [comment]
@philipp-sontag-by would recommend raising a new issue to track this. In fact we might want to do that with any of the meatier tasks left in this PR (if any)

**Comment by Matt711** [comment]
@philipp-sontag-by I think we can do this, especially since the point of the API is to cover most of the scheduler's methods. I'll create a follow-up PR to add this if @jacobtomlinson is okay with it.

**Comment by jacobtomlinson** [comment]
Yeah let's discuss this in a new issue. 

I'm curious what shutdown logic you want to do that retire workers isn't suitable for? As I see it the best way to gracefully shutdown is to allow the scheduler to end the worker process and then clean up the completed pods after.

**Comment by gjoseph92** [comment]
I'm concerned about a security regression here. By default, this is opening up an API that allows anyone to change cluster state (via `retire_workers` currently, but I imagine other things might be added someday too).

Prior to this, the only way to do things that affected cluster state was through the client. All the HTTP routes were effectively read-only. (Whether there is a vulnerability in the bokeh dashboard is another topic; it's pretty possible there is, but I'm just talking here in principle.)

I think it's rather common to expose the HTTP routes to the public internet. For example, I believe dask-cloudprovider does this:
> By default a Dask security group will be created with ports 8786 and 8787 exposed to the internet https://cloudprovider.dask.org/en/latest/aws.html#dask_cloudprovider.aws.EC2Cluster

You want those ports exposed for convenience, so you can connect to them. But you don't want anyone to be able to do stuff to the cluster, so you set up TLS using temporary credentials. dask-cloudprovider does this for you as well:
> When a cluster is launched with any of these cluster managers a set of temporary keys will be generated and distributed to the cluster nodes via their startup script. All communication between the client, scheduler and workers will then be encrypted and only clients and workers with valid certificates will be able to connect to the scheduler.
https://cloudprovider.dask.org/en/latest/security.html#authentication-and-encryption

Currently, if you set up TLS for your cluster, this is [mTLS](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/), meaning the scheduler verifies the client's certificate ([docs](https://distributed.dask.org/en/stable/tls.html#security-policy), [code](https://github.com/dask/distributed/blob/9bb999d4b66670570d9d8f53d04501dec6a25e7e/distributed/security.py#L275-L276)). This serves as a form of authentication and authorization: if you've set up cluster security, you can only tell the scheduler to do things if you hold a valid certificate.

However, the HTTP routes have no authentication (they use standard TLS, not mTLS, because mTLS would be very inconvenient when you want to look at the dashboard with a web browser).

So after this change, someone who had gone to the trouble to set up mTLS for their cluster (or was using the defaults of their cluster deployment system) would, by default, have an unauthenticated endpoint running that allowed anyone with access to `:8787` (aka the dashboard) to affect cluster state.

----

I think we should do two things:
1. Short-term: disable the HTTP API if TLS is specified on the scheduler. This is a reasonable default.
2. Long-term: figure out a security posture and authentication for the HTTP API that's consistent with the security posture of other things that can affect cluster state (aka the client).

**Comment by jakirkham** [comment]
@gjoseph92 can you please file this as a new issue to make it easier to track?

**Pull request merged**
